std/meta/trait.zig

```diff
@@ -0,0 +1,495 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const mem = std.mem;
+const debug = std.debug;
+const warn = debug.warn;
+
+const meta = @import("index.zig");
+
+//This is necessary if we want to return generic functions directly because of how the
+// the type erasure works. see:  #1375
+fn traitFnWorkaround(comptime T: type) bool
+{
+    return false;
+}
+
+///
+
+pub const TraitFn = @typeOf(traitFnWorkaround);
+
+///
+
+//////Trait generators
+
+//Need TraitList because compiler can't do varargs at comptime yet
+pub const TraitList = []const TraitFn;
+pub fn multiTrait(comptime traits: TraitList) TraitFn
+{
+    const Closure = struct.
+    {
+        pub fn trait(comptime T: type) bool
+        {
+            comptime
```

Comment:
Does this `comptime` allow the function to be called without `comptime` at call site?

Reply:
I see this a lot. I'll only comment on the `comptime` here.

Reply:
If I remove that comptime block:
```
trait.zig:32:26: error: variable of type 'fn(type)var' must be const or comptime
            for(traits) |t|
                         ^
```

Even though `traits` is comptime. You see comptime blocks all over here because without them it becomes difficult for me to reason about what will and will not be interpreted as a runtime value.

Reply:
Pretty sure you can use an `inline for` for this. This will probably look better than a comptime block in all functions, maybe